### ISSUE TITLE
Moving to Dynamic Website | Features/Framework 

### ISSUE THREAD

**Issue by thisisayush** [asetalias/asetalias.github.io#35]
- Events Management
    - Events Listing and Creation
        - Title
        - Description
        - Speaker
    - Inbuilt Registration Forms Module for each Event [Abandon Google Forms]
    - Attendees Module with options to mark Going/Not Going [for end users]  and Attended [Admin Module for efficient management of ODs]
   - Direct Posting of event on Social Platforms automatically along with Mailing List
   - Linking of Google Photos Albums for Gallery
   - Team Allocation Module (efficient management)
- Faculty Module
    - Listing of All Attendees.
    - Listing of All Events
    - Contact details of speakers/members/attendees/team
- Facebook/Google+ Login API Integration
- Login/Signup via Email
- User Module
    - Event Registration
    - Profile
    - Github Profile
    - Events Attended (will help in deciding next carrier of ALiAS Culture)
- Forum (Q&A) [Public]

**Comment by AnimeshVerma17** [comment]
Interesting.........I can work on this.

**Comment by aniketmaithani** [comment]
What is the choice of framework ??? 
- [ ] Django
- [ ] Flask
- [ ] Other 

My Two Cents on "Django" . Though flask will also do. :) But rest is upto you guys because the learning curve will be bit steep as you'll progress with any framework especially Python based.

**Comment by AnimeshVerma17** [comment]
Mine is PHP.

**Comment by AnimeshVerma17** [comment]
And BTW my PR(#33) is still pending, if anyone can merge it.

**Comment by aniketmaithani** [comment]
@AnimeshVerma17 this thread is about Feature/Future dev of Asetalias website. And you're pointing a completely different issue in this! Please adhere to the guidelines and we can take  #33 issue on different thread. :) Hope you understand my point.

**Comment by AnimeshVerma17** [comment]
Yep got it :P .

**Comment by thisisayush** [comment]
@aniketmaithani I'll go with PHP as I want to learn more of it.
Also, just for knowledge, I have set up a test server at hostinger.in with domain asetalias.ga .Hosting account was created to see the resource limits and I have found they are well enough for development.

**Comment by Tanya-Jain** [comment]
Another desirable feature: email templates which gets its data automatically updated with change in event and then can be broadcasted to speakers, concerned faculties and audience

**Comment by AnimeshVerma17** [comment]
@Tanya-Jain great suggestion.....in short an automated mailing system :)

**Comment by aniketmaithani** [comment]
Okay. Can any one of you put this thread on mailing list also. Asking people about their opinion of the choice of framework we should use.

**Comment by shivamdiehard** [comment]
I would like to go with php

**Comment by thisisayush** [comment]
What framework are we going with then? @aniketmaithani

**Comment by aniketmaithani** [comment]
@thisisayush can you post this thread to mailing list as well. So that we can have more consolidated view. :)

**Comment by aniketmaithani** [comment]
@bhanuvrat @arcolife your take??

**Comment by arcolife** [comment]
If the aim is to do this https://github.com/asetalias/asetalias.github.io/issues/27#issuecomment-283015052, it really shouldn't matter. Any framework will do as long as we understand what role do the sub-components play in order to deserve their respective M/V/T[C] titles. 

I'm personally interested in learning more about Elixir https://twitter.com/arcolife/status/826643671940464640

And with that, we could:
- expand on adding demos on a `/interests` route, 
- implement research work, 
- contribute to industry level tech,.
- a /help_wanted route (or something similar), 
- FOSS work needing contributors (our own personal version of https://www.codetriage.com/ )

..and not just merely develop a website. This will keep the website alive with PRs and what not, once development phase is over.

TL;DR: whatever involves the concept of channels + scalable arch + message passing / queuing. Preferably organized into their respective micro-services / containers.

**Comment by thisisayush** [comment]
@aniketmaithani I haven't done it before nor do I have the mailing list. I would be glad to do it under guidance first time.

**Comment by thisisayush** [comment]
@arcolife Sorry, but I didn't understood anything 😅

**Comment by aniketmaithani** [comment]
@thisisayush read about what is mailing list : https://en.wikipedia.org/wiki/Mailing_list

ASET ALIAS has a mailing list : https://groups.google.com/forum/#!forum/amity-linux-group

You post the mail here and it will be delivered to all the members :). DO adhere to the mailing list guidelines

**Comment by aniketmaithani** [comment]
@arcolife let's start with the basics. These are undergrads 1st year students :P

**Comment by thisisayush** [comment]
After thinking, I am in for any other framework than PHP, I suddenly remembered I had to build one project on PHP so I can go with other Platform for this one. Flask/Django as suggested would be good.
Other than that, MySQL could be used for database.
@aniketmaithani @ATB00ker @Tanya-Jain

**Comment by ATB00ker** [comment]
After looking at GSoC Organisations of this year... I am willing to work on python and Django. :)
P.S: MySQL works for me. :)

**Comment by ATB00ker** [comment]
Also, about the front-end of the website. Tanya is working on new color scheme, I have seen some screenshot and  it looks good, I think we can mordify the current front-end and adapt it for the dynamic site in future. :)

**Comment by AnimeshVerma17** [comment]
So what`s the final choice?

**Comment by ATB00ker** [comment]
P.S: My opinion is Django; It is based on python so we could start with basic python coding and use the high-level libraries of django as we are comfortable with the basics...

**Comment by AnimeshVerma17** [comment]
Well my first opinion is for PHP 7, if anyone wanna consider it once.
Nevertheless majority vote is most welcomed.

**Comment by thisisayush** [comment]
A step towards Dynamic #63

**Comment by bhanuvrat** [comment]
While it is a good idea, I wouldn't endorse making a web application only meant for asetalias.
Reasons: 
  - it takes too long to get one up and running.
  - most people will loose their steam by the time its done.
  - there would be only a handful of people using this application.
  - rarely someone outside Amity would be served by it.
  - whats the ****ing point?

# Drawbacks of starting a project from scratch:
  - as beginners you are more likely to learn bad practices.
  - you wouldn't have access to good code.
  - there is only so much that Aniket and Archit would be able to review.
  - if you are also learning in the process, will be too long before it finishes.

# Not in favour of PHP.
  - more often than not teaches you bad practices: all variables are global ! WTF?
  - skill is not appreciated financially: even if you get good, you aren't paid well.
  - too many sub-standard PHP developers out there, willing to work for way to less.
  - does not scale well.

# Choice of tools / language / framework:
- ensure there are good and experienced mentors for the technology around you.
- solicit their commitment to helping you with your problems, before you start.
- it should teach you something new: be it a new paradigm(functional programming wink wink) or better coding practices.
- it should neither be something very new (unstable) nor be very old (read cobol!)
etc ... 

# Suggestion

Make a generic application which can be deployed for diverse scenarios. 
Even better if you work on small reusable modules instead of a monolithic web app only meant for asetalias.

OR 

Use an existing application and modify it to do your bidding. This will teach you the most wanted skill: *understanding other people's code*.


# Bottom line:
Contact Sanyam Khurana and Amit Kumar from PyDelhi. They were working on confluence: aims to do similar things like you.
See if you can contribute to the same application and deploy it for ALiAS as well.
This application is in Django.


# Gyan
Generalize the problem statement, and make solutions that more people than just you would use.
Something that others will also have a vested interest in maintaining and keeping alive, even after you have moved on.

**Comment by thisisayush** [comment]
@bhanuvrat Bhaiya, The aim is to create a application dynamic enough to be implemented to variety of systems. I haven't worked with PHP or any other platform enough but seeing most simple web applications are running in PHP including the most popular CMS WordPress (though it has security issues) I don't think PHP has enough issues that can't be resolved via simple tricks (like SQL Injection can be avoided using manual filter or using prepared statements) but the platform is not much of important as SQL Queries still remain same, the algorithm still remains same, the data returned by server still remains same and the way login sessions work will still remain same. The aim is not to develop things but to learn the way to code or how things work. (Implementing JSON is to make the site backend ready if we go for it so that frontend may not require changes and can work as it is)
I don't know how other applications work, either they directly use PHP code in the page, I rely on AJAX Requests to php code for getting, updating and/or validate data and requests while on a php header to check login. (like <?php if(!isUserLoggedIn()){header("login.php")}?> )
The project may seem to be large, but those interested can work on specific modules like frontends for different pages, writing functions to database connection, writing login scripts, writing Javascripts, etc. This will provide knowledge enough on how things work, I don't expect Aniket or Archit bhaiya to review all code but only enough to tell us any of the commonly used practices.
I would love to go with DJango but not everyone (including me) would be comfortable with using Python and then Django would be just another level and hence as you said "most people will loose their steam by the time its done"

**Comment by bhanuvrat** [comment]
TL;DR

- please don't start a new project only for asetalias.
- there already is an app called confluence being worked on by Sanyam and Amit.
- We shall fork this application on sunday.

- starting with PHP will probably teach you anti-patterns: don't believe me? remind me to schedule a call with @amanjain (https://github.com/amanjain)
- work on things that more people will use. Trust me that will give you more satisfaction.
- reuse as much as you can, making a web-app for alias entirely on your own will not be as huge an achievement as you think it would be.


Now the detailed inline response to @thisisayush. 

    @bhanuvrat Bhaiya, 
Ayush dear
no bhaiya
no sire!
Anuvrat here.

    The aim is to create a application dynamic enough to be implemented to variety of systems.
Are there people ready to use your product?
Who are they?
By when would you be able to deliver a feature complete bug free product to them?

    I haven't worked with PHP or any other platform enough but seeing most simple web applications are running in PHP including the most popular CMS WordPress (though it has security issues) I don't think PHP has enough issues that can't be resolved via simple tricks

do you want to learn how things work or the tricks to make horrendous things work in a sane way?
Start with things that make you a better programmer, and then move on to murky areas so that your base is solid and you know the difference between good and bad practices.

Children first need to be taught "respect women" by their parents at home. That task can't be outsourced to the society at large.
Similarly, budding developers first need to learn that global variables are amazingly stupid idea ...


    (like SQL Injection can be avoided using manual filter or using prepared statements) but the platform is not much of important as SQL Queries still remain same, the algorithm still remains same, the data returned by server still remains same and the way login sessions work will still remain same.
Agreed, though not entirely.

    The aim is not to develop things but to learn the way to code or how things work.
So another college project that no one other than you and a clueless professor would ever use?
Would you want to write 10 lines of code which is used by 1000 people or 10,000 lines of code used only by 10 humans?


    (Implementing JSON is to make the site backend ready if we go for it so that frontend may not require changes and can work as it is)
    I don't know how other applications work, either they directly use PHP code in the page, I rely on AJAX Requests to php code for getting, updating and/or validate data and requests while on a php header to check login. (like <?php if(!isUserLoggedIn()){header("login.php")}?>)
Good thinking. Then start by designing and defining the workflow, APIs, and various screens.
Can the discussion on choice of technology for now.

    The project may seem to be large, 
It is humungous. To get a reasonably feature-complete bug-free version up and running on production would take experienced developers working full time around 2 months.
Students stealing time between classes, assignments, exams and dates: who knows how long.

    but those interested can work on specific modules like frontends for different pages, writing functions to database connection, writing login scripts, writing Javascripts, etc. 
The whole team would have to be equally dedicated, just you would not be enough and more likely to burn-out and abandon all hopes.

    This will provide knowledge enough on how things work, 
Maybe. "how things work" has a much broader scope. deploying, maintaining, bug fixing.
Your view currently is limited to developing the application, and you will only be able to learn the devops aspect once you complete the application and deploy it ... which would take forever.

    I don't expect Aniket or Archit bhaiya to review all code but only enough to tell us any of the commonly used practices.
you should. What is the point of working on something if no one is giving you constructive feedback on it?
Collaborating a small module on a project that experienced developers are contributing to will certainly give you that feedback.
Without which you would not know your blunders from your accomplishments.


    I would love to go with DJango but not everyone (including me) would be comfortable with using Python and then Django would be just another level and hence as you said "most people will loose their steam by the time its done" 
which is why I also implored everyone not to start a project from scratch. 
build on top of something that already exists (pydelhi's confluence _hint-hint-wink-wink_)
setting small targets and achieving them is much more feasible, productive and pleasurable than planning huge and not being able to achive it.

    and also on a bit research I found that using django or any other platform will require a VPS or Dedicated Server, just saying...

I shall give you two vps for your staging and production enviroments ... aur bolo?
